Lib/importlib/_bootstrap_external.py

```diff
@@ -93,35 +93,75 @@ def _unpack_uint16(data):
 
 
 if _MS_WINDOWS:
-    def _path_join(*path_parts):
-        """Replacement for os.path.join()."""
-        if not path_parts:
-            return ""
-        if len(path_parts) == 1:
-            return path_parts[0]
-        root = ""
-        path = []
-        for new_root, tail in map(_os._path_splitroot, path_parts):
-            if new_root.startswith(path_sep_tuple) or new_root.endswith(path_sep_tuple):
-                root = new_root.rstrip(path_separators) or root
-                path = [path_sep + tail]
-            elif new_root.endswith(':'):
-                if root.casefold() != new_root.casefold():
-                    # Drive relative paths have to be resolved by the OS, so we reset the
-                    # tail but do not add a path_sep prefix.
-                    root = new_root
-                    path = [tail]
-                else:
-                    path.append(tail)
+    def _path_splitroot(p):
+        """Replacement for os.path.splitroot()."""
+        sep = '\\'
+        altsep = '/'
+        colon = ':'
+        unc_prefix = '\\\\?\\UNC\\'
+        empty = ''
+        normp = p.replace(altsep, sep)
+        if normp[:1] == sep:
+            if normp[1:2] == sep:
+                # UNC drives, e.g. \\server\share or \\?\UNC\server\share
+                # Device drives, e.g. \\.\device or \\?\device
+                start = 8 if normp[:8].upper() == unc_prefix else 2
+                index = normp.find(sep, start)
+                if index == -1:
+                    return p, empty, empty
+                index2 = normp.find(sep, index + 1)
+                if index2 == -1:
+                    return p, empty, empty
+                return p[:index2], p[index2:index2 + 1], p[index2 + 1:]
+            else:
+                # Relative path with root, e.g. \Windows
+                return empty, p[:1], p[1:]
+        elif normp[1:2] == colon:
+            if normp[2:3] == sep:
+                # Absolute drive-letter path, e.g. X:\Windows
+                return p[:2], p[2:3], p[3:]
             else:
-                root = new_root or root
-                path.append(tail)
-        path = [p.rstrip(path_separators) for p in path if p]
-        if len(path) == 1 and not path[0]:
-            # Avoid losing the root's trailing separator when joining with nothing
-            return root + path_sep
-        return root + path_sep.join(path)
+                # Relative path with drive, e.g. X:Windows
+                return p[:2], empty, p[2:]
+        else:
+            # Relative path, e.g. Windows
+            return empty, empty, p
 
+    def _path_join(path, *paths):
```

Comment:
Maintenance, mainly. This code is changed frequently because it doesn't handle a lot of cases, and we treat the OS function as correct (with only a few exceptions that we already handled better). Usually on Windows we never hit this implementation, and so it's only for people on other platforms who deliberately import `ntpath`.

But I missed why `PathCchSkipRoot` (presumably) is missing? It says it should be available in UWP apps.

Reply:
The API functions declared in "pathcch.h" are in the app and system API partitions, but not the games partition, i.e.

```c
#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP | WINAPI_PARTITION_SYSTEM)
```

Thus they aren't available on Xbox.

> Usually on Windows we never hit this implementation, and so it's only for people on other platforms who deliberately import ntpath.

I may be wrong, but I think the question was about why there's a separate implementation of `os.path.join()` as `_path_join()` in `_bootstrap_external`.

As to `nt._path_splitroot()`, I hope that it will eventually be rewritten as an accelerated implementation of `ntpath.splitroot()`. The latter splits a path into 3 parts: drive, root, and the rest. This is the basis for several functions in `ntpath`, such as `join()`.

Reply:
Yes it is unavailable on the xbox. When the join method is irrelevant, I will revert this part. I only realized it failing in `_path_isabs` and thought probably both are required.

Reply:
edit: just tested and in our build we hit `_path_join` as well:
```
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 674, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible
  File "<frozen importlib._bootstrap_external>", line 1722, in _fix_up_module
  File "<frozen importlib._bootstrap_external>", line 805, in spec_from_file_location
  File "<frozen importlib._bootstrap_external>", line 104, in _path_join
AttributeError: module 'nt' has no attribute '_path_splitroot'
```